File: frontend/src/types/portfolio/transaction.ts
export type TransactionType = 'BUY' | 'SELL'

export interface Transaction {
  id: string
  assetId: string
  portfolioId: string
  type: TransactionType
  quantity: string // string — BigDecimal
  price: string // string — BigDecimal
  currency: string
  fee?: string | null
  executedAt: string
  notes?: string | null
}

export function isTransaction(obj: unknown): obj is Transaction {
  if (obj === null || typeof obj !== 'object') return false
  const candidate = obj as Record<string, unknown>

  return (
    typeof candidate.id === 'string' &&
    typeof candidate.assetId === 'string' &&
    typeof candidate.portfolioId === 'string' &&
    ['BUY', 'SELL'].includes(candidate.type as string) &&
    typeof candidate.quantity === 'string' &&
    typeof candidate.price === 'string' &&
    typeof candidate.currency === 'string' &&
    typeof candidate.executedAt === 'string'
  )
}

File: backend/services/portfolio-manager/src/main/java/com/fininsight/portfoliomanager/domain/Transaction.java
package com.fininsight.portfoliomanager.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import com.fininsight.portfoliomanager.domain.enums.TransactionType;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

@Entity(name = "DataTransaction")
@Table(name = "transactions")
@Getter
@Setter
@NoArgsConstructor
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class Transaction {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @EqualsAndHashCode.Include
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "asset_id")
    private Asset asset;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "portfolio_id", nullable = false)
    private Portfolio portfolio;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TransactionType type;

    @NotNull
    @Column(nullable = false, precision = 18, scale = 8)
    private BigDecimal quantity;

    @NotNull
    @Column(nullable = false, precision = 18, scale = 4)
    private BigDecimal price;

    @NotBlank
    @Column(nullable = false, length = 10)
    private String currency;

    @Column(precision = 18, scale = 4)
    private BigDecimal fee;

    @NotNull
    @Column(name = "executed_at", nullable = false)
    private Instant executedAt;

    @Column
    private String notes;
}
